storage/postgres/client.go

```diff
@@ -96,89 +95,99 @@ func (c *Client) SendBatch(ctx context.Context, batch *storage.QueryBatch) error
+	}
+	defer common.CloseOrLog(batchResults, c.logger)
+
+	// Exec indiviual queries in the batch.
+	for i := 0; i < pgxBatch.Len(); i++ {
+		if _, err := batchResults.Exec(); err != nil {
+			rollbackErr := ""
+			if useExplicitTx {
+				err2 := tx.Rollback(ctx)
+				if err2 != nil {
+					rollbackErr = fmt.Sprintf("; also failed to rollback tx: %s", err2.Error())
 				}
-				return fmt.Errorf("query %d %v: %w%s", i, batch.Queries()[i], err, rollbackErr)
 			}
+			return fmt.Errorf("query %d %v: %w%s", i, batch.Queries()[i], err, rollbackErr)
 		}
+	}
 
-		// Commit the tx.
-		if useExplicitTx {
-			err := tx.Commit(ctx)
-			if err != nil {
-				return fmt.Errorf("failed to commit tx: %w", err)
-			}
-		}
-	} else {
-		// Begin a transaction.
-		tx, err := c.pool.BeginTx(ctx, opts)
+	// Commit the tx.
+	if useExplicitTx {
+		err := tx.Commit(ctx)
 		if err != nil {
-			return fmt.Errorf("failed to begin tx: %w", err)
+			return fmt.Errorf("failed to commit tx: %w", err)
 		}
+	}
+	return nil
+}
 
-		// Exec indiviual queries in the batch.
-		for i, q := range batch.Queries() {
-			if _, err2 := tx.Exec(ctx, q.Cmd, q.Args...); err2 != nil {
-				rollbackErr := ""
-				err3 := tx.Rollback(ctx)
-				if err3 != nil {
-					rollbackErr = fmt.Sprintf("; also failed to rollback tx: %s", err3.Error())
-				}
-				return fmt.Errorf("query %d %v: %w%s", i, q, err2, rollbackErr)
+// Submits a new batch of queries, sending one query at a time. Compared with `sendBatchWithOptionsSlow`, this
+// gives slower performance but better error reporting.
+func (c *Client) sendBatchWithOptionsSlow(ctx context.Context, batch *storage.QueryBatch, opts pgx.TxOptions) error {
+	// Begin a transaction.
+	tx, err := c.pool.BeginTx(ctx, opts)
+	if err != nil {
+		return fmt.Errorf("failed to begin tx: %w", err)
+	}
+
+	// Exec indiviual queries in the batch.
+	for i, q := range batch.Queries() {
+		if _, err2 := tx.Exec(ctx, q.Cmd, q.Args...); err2 != nil {
+			rollbackErr := ""
+			err3 := tx.Rollback(ctx)
+			if err3 != nil {
+				rollbackErr = fmt.Sprintf("; also failed to rollback tx: %s", err3.Error())
 			}
+			return fmt.Errorf("query %d %v: %w%s", i, q, err2, rollbackErr)
 		}
+	}
 
-		// Commit the transaction.
-		err = tx.Commit(ctx)
-		if err != nil {
-			c.logger.Error("failed to submit tx",
-				"error", err,
-				"batch", batch.Queries(),
-			)
-			return err
-		}
+	// Commit the transaction.
+	err = tx.Commit(ctx)
+	if err != nil {
+		c.logger.Error("failed to submit tx",
+			"error", err,
+			"batch", batch.Queries(),
+		)
+		return err
 	}
 	return nil
 }
 
+func (c *Client) SendBatchWithOptions(ctx context.Context, batch *storage.QueryBatch, opts pgx.TxOptions) error {
+	if err := c.sendBatchWithOptionsFast(ctx, batch, opts); err == nil {
+		// The fast path succeeded. This should happen most of the time.
+		return nil
+	}
+	// There was an error. The tx was reverted, so we can resubmit. This time, use the slow method for better error msgs.
```

Comment:
> was reverted

or there was an error in `tx.Commit()`, and we don't know :grimacing:

Reply:
Hm ... that is an interesting corner case.

If there was a problem _sending_ the `tx.Commit()`, or the commit operation somehow errored out, we're fine because the tx indeed never committed.
If there was a problem _receiving the results_ the `tx.Commit()`, we're in trouble. However the only case where I can think of this happening is with connectivity problems. If the db was unreachable _just for the commit_ (and not when sending the batch), it seems extremely unlikely for it to be suddenly reachable again in the next split second when resending the tx.

However I'm inclined to risk it, because **no aspect of this problem is specific to the retry logic in this PR**. If pgx reports an error when sending the tx, we (and presumably most other clients) always assume the tx had no side effects. I don't see a way around it other than some sort of two-phase commits.

Also, specific to nexus: If a tx succeeds but we don't learn about it so we then erroneously re-send the same tx, in many cases the doubled tx will cause PK violations, so that's another failsafe.

Reply:
pk violations :pray: 

although we're also moving more to upsert style in various places

Reply:
ok so from other threads, there's no application layer database transaction, so we can resubmit